xapian-core/include/xapian/weight.h

```diff
@@ -590,6 +590,118 @@ class XAPIAN_VISIBILITY_DEFAULT BM25Weight : public Weight {
+     *		   wdf influences the weights.  (default 1)
+     *
+     *  @param k3  A non-negative parameter controlling how influential
+     *		   within-query-frequency (wqf) is.  k3=0 means that wqf
+     *		   doesn't affect the weights.  The larger k3 is, the more
+     *		   wqf influences the weights.  (default 1)
+     *
+     *  @param b   A parameter between 0 and 1, controlling how strong the
+     *		   document length normalisation of wdf is.  0 means no
+     *		   normalisation; 1 means full normalisation.  (default 0.5)
+     *
+     *  @param min_normlen  A parameter specifying a minimum value for
+     *		   normalised document length.  Normalised document length
+     *		   values less than this will be clamped to this value, helping
+     *		   to prevent very short documents getting large weights.
+     *		   (default 0.5)
+     *
+     *  @param delta  A parameter for pseudo tf value to control the scale
+     *		      of the tf lower bound. Delta(δ) can be tuned for example
+     *		      from 0.0 to 1.5 but BM25+ can still work effectively
+     *		      across collections with a fixed δ = 1.0. (default 1.0)
+     */
+    BM25PlusWeight(double k1, double k3, double b, double min_normlen, double delta)
+	: param_k1(k1), param_k3(k3), param_b(b), param_min_normlen(min_normlen),
+	  param_delta(delta)
+    {
+	if (param_k1 < 0) param_k1 = 0;
+	if (param_k3 < 0) param_k3 = 0;
+	if (param_delta < 0) param_delta = 0;
+	if (param_b < 0) {
+	    param_b = 0;
+	} else if (param_b > 1) {
+	    param_b = 1;
+	}
+	need_stat(COLLECTION_SIZE);
+	need_stat(RSET_SIZE);
+	need_stat(TERMFREQ);
+	need_stat(RELTERMFREQ);
+	need_stat(WDF);
+	need_stat(WDF_MAX);
+	if (param_k1 != 0 && param_b != 0) {
+	    need_stat(DOC_LENGTH_MIN);
+	    need_stat(AVERAGE_LENGTH);
+	}
+	if (param_k1 != 0 && param_b != 0) need_stat(DOC_LENGTH);
+	if (param_k3 != 0) need_stat(WQF);
+	if (param_delta != 0) {
+	    need_stat(AVERAGE_LENGTH);
+	    need_stat(DOC_LENGTH);
+	    need_stat(WQF);
+	}
+    }
+
+    BM25PlusWeight()
+	: param_k1(1), param_k3(1), param_b(0.5), param_min_normlen(0.5),
+	  param_delta(1)
+    {
+	need_stat(COLLECTION_SIZE);
+	need_stat(RSET_SIZE);
+	need_stat(TERMFREQ);
+	need_stat(RELTERMFREQ);
+	need_stat(WDF);
+	need_stat(WDF_MAX);
+	need_stat(DOC_LENGTH_MIN);
+	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(WQF);
+    }
+
+    std::string name() const;
+
+    std::string serialise() const;
+    BM25PlusWeight * unserialise(const std::string & serialised) const;
+
+    double get_sumpart(Xapian::termcount wdf,
+		       Xapian::termcount doclen,
+		       Xapian::termcount uniqterm) const;
+    double get_maxpart() const;
+
+    double get_sumextra(Xapian::termcount doclen,
+			Xapian::termcount uniqterms) const;
+    double get_maxextra() const;
+
+  protected:
+    /// Factor to multiply the document length by.
+    mutable Xapian::doclength len_factor;
```

Comment:
I've been working on a different machine lately that didn't have valgrind installed until now; just installed it. Sorry, will be careful now.